File: src/csharp/Intel/Iced.UnitTests/Intel/DecoderTests/DecoderTestCases.cs
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

using System;
using System.Linq;

namespace Iced.UnitTests.Intel.DecoderTests {
	static class DecoderTestCases {
		public static readonly DecoderTestCase[] TestCases16 = ReadTestCases(16);
		public static readonly DecoderTestCase[] TestCases32 = ReadTestCases(32);
		public static readonly DecoderTestCase[] TestCases64 = ReadTestCases(64);
		public static readonly DecoderTestCase[] TestCasesMisc16 = ReadMiscTestCases(16);
		public static readonly DecoderTestCase[] TestCasesMisc32 = ReadMiscTestCases(32);
		public static readonly DecoderTestCase[] TestCasesMisc64 = ReadMiscTestCases(64);
		public static readonly DecoderMemoryTestCase[] TestCasesMemory16 = ReadMemoryTestCases(16);
		public static readonly DecoderMemoryTestCase[] TestCasesMemory32 = ReadMemoryTestCases(32);
		public static readonly DecoderMemoryTestCase[] TestCasesMemory64 = ReadMemoryTestCases(64);

		public static DecoderTestCase[] GetTestCases(int bitness) =>
			bitness switch {
				16 => TestCases16,
				32 => TestCases32,
				64 => TestCases64,
				_ => throw new ArgumentOutOfRangeException(nameof(bitness)),
			};

		public static DecoderTestCase[] GetMiscTestCases(int bitness) =>
			bitness switch {
				16 => TestCasesMisc16,
				32 => TestCasesMisc32,
				64 => TestCasesMisc64,
				_ => throw new ArgumentOutOfRangeException(nameof(bitness)),
			};

		public static DecoderMemoryTestCase[] GetMemoryTestCases(int bitness) =>
			bitness switch {
				16 => TestCasesMemory16,
				32 => TestCasesMemory32,
				64 => TestCasesMemory64,
				_ => throw new ArgumentOutOfRangeException(nameof(bitness)),
			};

		static DecoderTestCase[] ReadTestCases(int bitness) {
			var filename = PathUtils.GetTestTextFilename($"DecoderTest{bitness}.txt", "Decoder");
			return DecoderTestParser.ReadFile(bitness, filename).ToArray();
		}

		static DecoderTestCase[] ReadMiscTestCases(int bitness) {
			var filename = PathUtils.GetTestTextFilename($"DecoderTestMisc{bitness}.txt", "Decoder");
			return DecoderTestParser.ReadFile(bitness, filename).ToArray();
		}

		static DecoderMemoryTestCase[] ReadMemoryTestCases(int bitness) {
			var filename = PathUtils.GetTestTextFilename($"MemoryTest{bitness}.txt", "Decoder");
			return MemoryDecoderTestParser.ReadFile(bitness, filename).ToArray();
		}
	}
}

File: src/java/iced-x86/src/test/java/com/github/icedland/iced/x86/dec/DecoderTestCases.java
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

package com.github.icedland.iced.x86.dec;

import com.github.icedland.iced.x86.PathUtils;

final class DecoderTestCases {
	public static final DecoderTestCase[] testCases16;
	public static final DecoderTestCase[] testCases32;
	public static final DecoderTestCase[] testCases64;
	public static final DecoderTestCase[] testCasesMisc16;
	public static final DecoderTestCase[] testCasesMisc32;
	public static final DecoderTestCase[] testCasesMisc64;
	public static final DecoderMemoryTestCase[] testCasesMemory16;
	public static final DecoderMemoryTestCase[] testCasesMemory32;
	public static final DecoderMemoryTestCase[] testCasesMemory64;

	static {
		testCases16 = readTestCases(16);
		testCases32 = readTestCases(32);
		testCases64 = readTestCases(64);
		testCasesMisc16 = readMiscTestCases(16);
		testCasesMisc32 = readMiscTestCases(32);
		testCasesMisc64 = readMiscTestCases(64);
		testCasesMemory16 = readMemoryTestCases(16);
		testCasesMemory32 = readMemoryTestCases(32);
		testCasesMemory64 = readMemoryTestCases(64);
	}

	public static DecoderTestCase[] getTestCases(int bitness) {
		switch (bitness) {
		case 16:
			return testCases16;
		case 32:
			return testCases32;
		case 64:
			return testCases64;
		default:
			throw new UnsupportedOperationException();
		}
	}

	public static DecoderTestCase[] getMiscTestCases(int bitness) {
		switch (bitness) {
		case 16:
			return testCasesMisc16;
		case 32:
			return testCasesMisc32;
		case 64:
			return testCasesMisc64;
		default:
			throw new UnsupportedOperationException();
		}
	}

	public static DecoderMemoryTestCase[] getMemoryTestCases(int bitness) {
		switch (bitness) {
		case 16:
			return testCasesMemory16;
		case 32:
			return testCasesMemory32;
		case 64:
			return testCasesMemory64;
		default:
			throw new UnsupportedOperationException();
		}
	}

	private static DecoderTestCase[] readTestCases(int bitness) {
		String filename = PathUtils.getTestTextFilename("Decoder", String.format("DecoderTest%d.txt", bitness));
		return DecoderTestParser.readFile(bitness, filename);
	}

	private static DecoderTestCase[] readMiscTestCases(int bitness) {
		String filename = PathUtils.getTestTextFilename("Decoder", String.format("DecoderTestMisc%d.txt", bitness));
		return DecoderTestParser.readFile(bitness, filename);
	}

	private static DecoderMemoryTestCase[] readMemoryTestCases(int bitness) {
		String filename = PathUtils.getTestTextFilename("Decoder", String.format("MemoryTest%d.txt", bitness));
		return MemoryDecoderTestParser.readFile(bitness, filename);
	}
}
